sunpy/map/mapbase.py

```diff
@@ -562,71 +549,75 @@ def coordinate_system(self):
+        A list of supported coordinate systems.
 
-        if isinstance(carrington_longitude, str):
-            carrington_longitude = float(carrington_longitude)
+        This is a list so it can easily maintain a strict order. The list of
+        two element tuples, the first item in the tuple is the keys that need
+        to be in the header to use this coordinate system and the second is the
+        kwargs to SkyCoord.
+        """
+        return [(('hgln_obs', 'hglt_obs', 'dsun_obs'), {'lon': self.meta.get('hgln_obs'),
+                                                        'lat': self.meta.get('hglt_obs'),
+                                                        'radius': self.meta.get('dsun_obs'),
+                                                        'unit': (u.deg, u.deg, u.m),
+                                                        'frame': "heliographic_stonyhurst"}),
+                (('crln_obs', 'crlt_obs', 'dsun_obs'), {'lon': self.meta.get('crln_obs'),
+                                                        'lat': self.meta.get('crlt_obs'),
+                                                        'radius': self.meta.get('dsun_obs'),
+                                                        'unit': (u.deg, u.deg, u.m),
+                                                        'frame': "heliographic_carrington"}),]
 
-        return u.Quantity(carrington_longitude, 'deg')
+    def _remove_existing_observer_location(self):
+        """
+        Remove all keys that this map might use for observer location.
+        """
+        all_keys = expand_list([e[0] for e in self._supported_observer_coordinates])
+        for key in all_keys:
+            self.meta.pop(key)
 
     @property
-    def heliographic_latitude(self):
-        """Heliographic latitude."""
-        heliographic_latitude = self.meta.get('hglt_obs',
-                                              self.meta.get('crlt_obs',
-                                                            self.meta.get('solar_b0', None)))
+    def observer_coordinate(self):
+        """
+        The Heliographic Stonyhurst Coordinate of the observer.
+        """
+        for keys, kwargs in self._supported_observer_coordinates:
+            if all([k in self.meta for k in keys]):
+                return SkyCoord(obstime=self.date, **kwargs).heliographic_stonyhurst
 
-        if heliographic_latitude is None:
-            if self._default_heliographic_latitude is None:
-                warnings.warn("Missing metadata for heliographic latitude: assuming Earth-based observer.",
-                              SunpyUserWarning)
-                self._default_heliographic_latitude = get_sun_B0(self.date)
-            heliographic_latitude = self._default_heliographic_latitude
+        all_keys = [str(e[0]) for e in self._supported_observer_coordinates]
+        all_keys = '\n'.join(all_keys)
+        warning_message = ("Missing metadata for observer: assuming Earth-based observer."
```

Comment:
I think it would be useful to tell the user exactly which metadata is missing here (instead of just telling them that something is missing).

Reply:
It's not trivial to work out what ones are missing, i.e. you could have a weird header with a mix of crln and hglt and it would be hard to work out which one the user was aiming for.

I will give it some thought.

Reply:
I think improved docs on map meta will really help here

Reply:
I opened an issue to keep track of this because I didn't want to block this on it.